src/test/java/api/support/fakes/FakePubSub.java

```diff
@@ -42,8 +42,8 @@ public static void register(Router router) {
             .setStatusCode(HTTP_BAD_REQUEST.toInt())
             .putHeader("content-type", "text/plain; charset=utf-8")
             .putHeader("content-length", Integer.toString(buffer.length()))
-            .write(buffer)
-            .end();
+            .write(buffer);
+          routingContext.response().end();
```

Comment:
Why can `end` no longer be chained?

Reply:
In vertx 4, both the end() and write() methods are no longer fluent, which means they now return either void or future<void> depending on where they are used. This is a breaking change according to the upgrade docs.  In practical terms, this means trying to chain off either method will now generate a compilation error (which is what was happening here).

Reply:
Thanks, I didn't realise that `write` was no longer fluent. I wonder why they did that.